### ISSUE TITLE
Migrate AdSense Top Earning Pages dashboard widget to use Widgets API

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#1902]
The current AdSense `DashboardAdSenseTopEarningPagesSmall` widget should be migrated to use the Widgets API (which is behind a feature flag), while maintaining the current version as legacy components. Those will be removed only once all widgets have been migrated so that we can do the entire switch.

Some general practices for migrating:
* The migration should happen by fully **re-implementing** the components: In other words, the existing components should keep their current functionality, they should **not** use the datastore. Those legacy components should be prefixed with `Legacy*`.
* A good starting point is probably to copy the components into new files and go from there, while renaming the old ones and leaving them untouched.
* For the refactored version, _only_ the actual widget components should be fully responsible for retrieving data from the datastore. If they rely on further `common` components for UI, these should receive such data via props.
* If there are UI parts that can be reused between the new and legacy components, those can be outsourced into `common` components, however as mentioned above **it is important that the legacy components keep using the legacy data API and are not switched to the datastore**. Only the new ones should be.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The [`DashboardAdSenseTopEarningPagesSmall`](https://github.com/google/site-kit-wp/blob/1.13.1/assets/js/modules/adsense/components/dashboard/DashboardAdSenseTopEarningPagesSmall.js) and [`AdSenseDashboardWidgetTopPagesTableSmall`](https://github.com/google/site-kit-wp/blob/1.13.1/assets/js/modules/analytics/components/dashboard/AdSenseDashboardWidgetTopPagesTableSmall.js) components should be migrated to a single functional component using the datastore entirely, no globals and no legacy data API.
* Refactoring map:
    * `dashboard/DashboardAdSenseTopEarningPagesSmall.js` and the Analytics `dashboard/AdSenseDashboardWidgetTopPagesTableSmall.js` components should be combined into a single `dashboard/DashboardTopEarningPagesWidget.js` (refactored), plus `dashboard/LegacyDashboardAdSenseTopEarningPagesSmall.js` (legacy 1) and Analytics `dashboard/LegacyAdSenseDashboardWidgetTopPagesTableSmall.js` (legacy 2). The refactored component should access the datastore for the main API data.
* The new `DashboardTopEarningPagesWidget` should perform checks via the datastore:
    * If the AdSense module is not active, return `null`.
    * If the Analytics module is not active, return `null`.
    * Only otherwise request and render actual data (via `modules/analytics` datastore), i.e. this probably needs to go into a separate component `DashboardTopEarningPages`, since React hooks cannot run conditionally (we use `*Widget` as top-level entry point while the un-suffixed version is essentially the main part of it, without any extra checks for active modules etc.).
    * `DashboardTopEarningPages` will need to specifically account for the API error where AdSense metrics are restricted, in the same way and with the same CTA message that the plugin shows today in such a case (to link AdSense and Analytics account).
* The new `DashboardTopEarningPagesWidget` component should be registered as a widget with priority 2, width `Widgets.WIDGET_WIDTHS.HALF`, and for the widget area `AREA_DASHBOARD_EARNINGS`.

#### Stage 1
* Rename the following files and components accordingly:
  * `assets/js/modules/adsence/dashboard/DashboardAdSenseTopEarningPagesSmall.js` -> `assets/js/modules/adsence/dashboard/LegacyDashboardAdSenseTopEarningPagesSmall.js`;
  * `assets/js/modules/analytics/dashboard/AdSenseDashboardWidgetTopPagesTableSmall.js` -> `assets/js/modules/analytics/dashboard/LegacyAdSenseDashboardWidgetTopPagesTableSmall.js`.
* Update `assets/js/modules/analytics/index.legacy.js` file to use `LegacyAdSenseDashboardWidgetTopPagesTableSmall` instead of `AdSenseDashboardWidgetTopPagesTableSmall`.

#### Stage 2
* Create `assets/js/modules/adsense/dashboard/DashboardTopEarningPagesWidget.js` file;
* `DashboardTopEarningPagesWidget.js` file should:
  * Define a new functional component with `DashboardTopEarningPagesWidget` name;
  * Use `useSelect` hook and the following selectors:
    * `getDateRange` selector from the `core/user` datastore to retrieve the current date range;
    * `getReport` selector from the analytics datastore with the following settings:
      ```js
      {
        dateRange: '...', // use the result from getDateRange selector
        dimensions: ['ga:pageTitle', 'ga:pagePath'],
        metrics: [
          { expression: 'ga:adsenseRevenue', alias: 'Earnings' },
          { expression: 'ga:adsenseECPM', alias: 'Page RPM' },
          { expression: 'ga:adsensePageImpressions', alias: 'Impressions' },
        ],
        orderby: [
          { fieldName: 'ga:adsenseRevenue', sortOrder: 'DESCENDING' },
        ],
        limit: 10,
      }
      ```
    * `getErrorForSelector` selector (added in the #1814) from the analytics datastore with `'getReport'` and 
      `[ argsUsedInGetReportSelector ]` arguments.
  * Render `AdSenseLinkCTA` if the error is not empty and the reason is `badRequest`, otherwise use just error message;
  * Replicate the logic from the legacy comonent to render top earning pages widget if the error is empty;
  * Wrap rendered component with the `Widget` component that has `adsenseTopEarningPages` slug  and `googlesitekit-widget--no-padding` class, and return it.
* Export a component enhanced with `whenActive` HOC (see Stage 2 in #1896) two times: one with `adsence` module name and one with `analytics` module name. This is needed to make sure we display this widget if both modules are active.

#### Stage 3
* Update `domReady` callback in the `assets/js/modules/adsense/index.js` file (added in the #1901):
  * Register `DashboardTopEarningPagesWidget` widget with the following arguments:
    * A string `adsenseTopEarningPages` as widget slug;
    * Settings object where:
      * `component` is `DashboardTopEarningPagesWidget`;
      * `width` is `Widgets.WIDGET_WIDTHS.HALF`;
      * `priority` is `2`;
      * `wrapWidget` is `false`.
    * Areas array that includes just `AREA_DASHBOARD_EARNINGS` imported from `googlesitekit/widgets/default-areas`.

## QA Brief

* In a development build, the new DashboardTopEarningPagesWidget widget should render correctly in the Site Kit dashboard. They should behave the same way as the old ones.
* It should be tested that they display errors if e.g. the API request goes wrong - they shouldn't keep loading infinitely in such a case (try e.g. to block client-to-server requests to cause an error).
* The old/legacy DashboardTopEarningPagesWidget widget should continue to show up in both development and production builds and behave like before.

## Changelog entry

* Migrate AdSense Top Earning Pages widget to new Widget API.

**Comment by felixarntz** [comment]
IB ✅

**Comment by felixarntz** [comment]
AdSense not linked state: ❌  incorrect because it should show the AdSense not linked CTA instead of this error
![Screenshot 2020-09-28 at 08 32 44](https://user-images.githubusercontent.com/3531426/94453435-833bef80-0165-11eb-8c3e-0e3af4469620.png)

AdSense linked state: ✅ 
![Screenshot 2020-09-28 at 08 31 18](https://user-images.githubusercontent.com/3531426/94453503-98b11980-0165-11eb-96e1-b06fc491ff71.png)

(Note how the widget boxes have different heights but that's okay. Will be fixed via #2022.)

QA ❌ , but it's okay to move forward with the release since it's not visible in the production bundle anyway.

**Comment by felixarntz** [comment]
The above remaining issue will be fixed via #2098.
